src/eh1/io.rs

```diff
@@ -0,0 +1,617 @@
+            expected_err: None,
+        }
+    }
+
+    /// Create a fill buffer transaction
+    pub fn fill_buf(response: Vec<u8>) -> Transaction {
+        Transaction {
+            expected_mode: Mode::FillBuff,
+            expected_data: Vec::new(),
+            response,
+            expected_err: None,
+        }
+    }
+
+    /// Create a consume transaction
+    pub fn consume(consumed: usize) -> Transaction {
+        Transaction {
+            expected_mode: Mode::Consume(consumed),
+            expected_data: Vec::new(),
+            response: Vec::new(),
+            expected_err: None,
+        }
+    }
+
+    /// Add an error return to a transaction
+    ///
+    /// This is used to mock failure behaviors.
+    ///
+    /// Note: When attaching this to a read transaction, the response in the
+    /// expectation will not actually be written to the buffer.
+    pub fn with_error(mut self, error: ErrorKind) -> Self {
+        self.expected_err = Some(error);
+        self
+    }
+}
+
+/// Mock IO implementation
+pub type Mock = Generic<Transaction>;
+
+impl ErrorType for Mock {
+    type Error = ErrorKind;
+}
+
+impl Write for Mock {
+    fn write(&mut self, buffer: &[u8]) -> Result<usize, Self::Error> {
+        let w = self.next().expect("no expectation for io::write call");
+        assert_eq!(w.expected_mode, Mode::Write, "io::write unexpected mode");
+        assert_eq!(
+            &w.expected_data, &buffer,
+            "io::write data does not match expectation"
+        );
+
+        match w.expected_err {
+            Some(err) => Err(err),
+            None => Ok(buffer.len()),
+        }
+    }
+
+    fn flush(&mut self) -> Result<(), Self::Error> {
+        let w = self.next().expect("no expectation for io::flush call");
+        assert_eq!(w.expected_mode, Mode::Flush, "io::flush unexpected mode");
+
+        match w.expected_err {
+            Some(err) => Err(err),
+            None => Ok(()),
+        }
+    }
+}
+
+impl Read for Mock {
+    fn read(&mut self, buffer: &mut [u8]) -> Result<usize, Self::Error> {
+        let w = self.next().expect("no expectation for io::read call");
+        assert_eq!(w.expected_mode, Mode::Read, "io::read unexpected mode");
+        buffer.copy_from_slice(&w.response);
+
+        match w.expected_err {
+            Some(err) => Err(err),
+            None => Ok(buffer.len()),
+        }
+    }
+}
+
+impl Seek for Mock {
+    fn seek(&mut self, pos: SeekFrom) -> Result<u64, Self::Error> {
+        let w = self.next().expect("no expectation for io::seek call");
+
+        if let Mode::Seek(expected_pos) = w.expected_mode {
+            assert_eq!(expected_pos, pos, "io::seek unexpected mode");
+
+            let ret_offset: u64 = u64::from_be_bytes(w.response.try_into().unwrap());
+            match w.expected_err {
+                Some(err) => Err(err),
+                None => Ok(ret_offset),
+            }
+        } else {
+            panic!("unexpected seek mode");
```

Comment:
As a user of the library, this can lead to hard-to-debug test failures.

This panic happens if the test expects a seek transaction, but some other action is triggered by the code.

Something like this might be much better:

```rust
panic!("expected seek transaction, but instead encountered {:?}", w.expected_mode);
```

(Same thing for all other impls.)